### ISSUE TITLE
Event Grid output binding does not work on Python

### ISSUE THREAD

**Issue by johndowns** [Azure/azure-functions-python-worker#638]
In the process of updating the [documentation for the Event Grid output binding](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-grid) (see [this PR](https://github.com/MicrosoftDocs/azure-docs-pr/pull/103060)), I discovered that the output binding does not work correctly with Python functions. It does work correctly for .NET and Node.js. (Note I have opened a [separate issue to log that this doesn't work for Java either](https://github.com/Azure/azure-functions-java-worker/issues/355).)

* I used this [function.json](https://gist.github.com/johndowns/e5bf2bf4a3bef84545dbd3e354215222) and this [function code](https://gist.github.com/johndowns/03be65daf9a37e460446d3b7e9f84883).
* When I ran the function, I got this error:

```
Result: Failure
Exception: FunctionLoadError: cannot load the MyFunction function: Python return annotation "EventGridEvent" does not match binding type "eventGrid"
Stack:   File "/azure-functions-host/workers/python/3.7/LINUX/X64/azure_functions_worker/dispatcher.py", line 245, in _handle__function_load_request
    function_id, func, func_request.metadata)
  File "/azure-functions-host/workers/python/3.7/LINUX/X64/azure_functions_worker/functions.py", line 241, in add_function
    f'Python return annotation "{return_pytype.__name__}" '
```

* I wonder if the issue is that the [binding extension bundle](https://github.com/Azure/azure-functions-extension-bundles/blob/master/src/Microsoft.Azure.Functions.ExtensionBundle/extensions.json#L53) does not include a reference to the `eventGrid` output binding (only to `eventGridTrigger`) - I have no idea how this works though, and if it's sufficient to simply put a PR adding that one line to the `extensions.json` or if something more needs to be done to wire it up.

**Comment by pragnagopa** [comment]
@Hazhzeng - can you please take a look? Also, add E2E test for event grid if not exists

**Comment by pragnagopa** [comment]
@johndowns - For Java, please see: https://github.com/Azure/azure-functions-java-worker/blob/dev/endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventGridTriggerTests.java for an example. If it still does not work, please open a separate issue here: https://github.com/Azure/azure-functions-java-worker/

**Comment by johndowns** [comment]
> @johndowns - For Java, please see: https://github.com/Azure/azure-functions-java-worker/blob/dev/endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventGridTriggerTests.java for an example. If it still does not work, please open a separate issue here: https://github.com/Azure/azure-functions-java-worker/

Thanks. That test doesn't invoke the output binding, just the trigger, so my original point stands. I will open a new issue on the other repo and adjust the description of this issue to restrict it to Python.

**Comment by Hazhzeng** [comment]
Yea, I think this issue is more of a feature request. According to the [documents](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-grid-output?tabs=python), Python and Java does not support Output Binding for Event Grid yet.

I will discuss with my team and update you once this is ready.

**Comment by johndowns** [comment]
@Hazhzeng yes, I wrote that part of the document :) I spoke to Jeff Hollan about this and he asked me to open this issue as the bindings should be consistent across languages.

**Comment by vrdmr** [comment]
Hi, @johndowns, Could you tell me a little bit more about the app configurations you used to test? Did all the three - python, nodejs and .NET had the same configuration (finding the difference between working and non-working scenarios)?
What was the Trigger? 
and what was the OutputBinding?

The questions above are mainly to clarify the setup, as I recently discovered multiple issues with the EventGrid extension -
1. https://github.com/Azure/azure-functions-templates/issues/942 - The issue is if the EventGrid extension is not picked up explicitly and eventGrid isn't the trigger, the following error shows up.
```
[4/20/20 6:19:35 PM] The 'HttpTrigger1' function is in error: The binding type(s) 'eventGrid' are 
not registered. Please ensure the type is correct and the binding extension is installed.
```
2. https://github.com/Azure/azure-functions-eventgrid-extension/issues/82: Even when 1 is fixed, there is an issue in host extension which has an issue with new netcore.
```
[4/20/20 7:23:53 PM] Executed 'Functions.HttpTrigger1' (Failed, Id=a0ae39ab-1e6a-49bf-b19b-d351aa7a9893)
[4/20/20 7:23:53 PM] System.Private.CoreLib: Exception while executing function: Functions.HttpTrigger1. Microsoft.Azure.WebJobs.Host: 
Unable to resolve the value for property 'EventGridAttribute.TopicEndpointUri'. 
Make sure the setting exists and has a valid value.
```

for python - I have added the EventGrid output binding now (https://github.com/Azure/azure-functions-python-library/pull/56) and it should be going out soon.

**Comment by vrdmr** [comment]
Hi, @johndowns,

The following example works without my fix (https://github.com/Azure/azure-functions-python-library/pull/56) on python. The extension needs to be loaded (due to the bug: https://github.com/Azure/azure-functions-templates/issues/942) but once you do that, the eventGrid custom topics would get populated.

```python
def main(eventGridEvent: func.EventGridEvent) -> str: 
    print("The function got the following event from EventGrid: ", eventGridEvent)

    output = {
    "id": "1111", 
    "subject": 'pythonapp/motorcycles', 
    "dataVersion": '1.0',
    "eventType": 'recordInserted', 
    "data": "event-data-test", 
    "eventTime": datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")}

    return json.dumps(output)
```

`function.json`
```json
{
  "scriptFile": "__init__.py",
  "bindings": [
    {
      "type": "eventGridTrigger",
      "name": "eventGridEvent",
      "direction": "in"
    },
    {
      "type": "eventGrid",
      "name": "$return",
      "topicEndpointUri": "MyEventGridTopicUriSetting",
      "topicKeySetting": "MyEventGridTopicKeySetting",
      "direction": "out"
    }
  ],
  "disabled": false
}
```

I'll try to give you an example with the new changes as well.

**Comment by vrdmr** [comment]
With https://github.com/Azure/azure-functions-python-library/pull/56 fix, the EventGrid output binding can be used, with `EventGridOutputEvent` needed to be returned. The change has been merged and we would be releasing it soon.

And example is below:
```python
def main(eventGridEvent: func.EventGridEvent, outputEvent: func.Out[func.EventGridOutputEvent]) -> None:
    logging.log("Triggering eventGridEvent: ", eventGridEvent)

    outputEvent.set(func.EventGridOutputEvent(id="1807",
                                      data={"make": "Ducati", "model": "Monster"},
                                      subject="pythonapp/motorcycles",
                                      event_type="recordInserted",
                                      event_time=datetime.datetime.utcnow(),
                                      data_version="1.0"))
```

```json
{
  "scriptFile": "__init__.py",
  "bindings": [
    {
      "type": "eventGridTrigger",
      "name": "eventGridEvent",
      "direction": "in"
    },
    {
      "type": "eventGrid",
      "name": "outputEvent",
      "topicEndpointUri": "MyEventGridTopicUriSetting",
      "topicKeySetting": "MyEventGridTopicKeySetting",
      "direction": "out"
    }
  ],
  "disabled": false
}
```

**Comment by johndowns** [comment]
Hi @vrdmr, I'm afraid it was a while ago that I tested this all so I don't have the details of exactly what I did. (I should also note that I'm not a Python developer so I'm not even sure I used the Python functions correctly - so I wouldn't put too much stock in my skills there!)

From memory though, I created an HTTP triggered function and had the Event Grid output binding, which didn't work for Python - hence logging this issue.

It's great to see this works now that you've made those changes.

Were you planning on updating [the documentation](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-grid-output?tabs=csharp) to include the Python examples?

**Comment by vrdmr** [comment]
@johndowns - I'll send in an update. 
Do I need to send in a PR to change this file - [functions-bindings-event-grid-output.md](https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/azure-functions/functions-bindings-event-grid-output.md).

**Comment by johndowns** [comment]
Yes, I believe that's the page. It's been moved since I worked on it but it looks like that's the one to update.

**Comment by vrdmr** [comment]
Thanks @johndowns. I've opened a documentation PR: https://github.com/MicrosoftDocs/azure-docs/pull/54037

I'll close this issue as it is resolved from our side and is now supported in the newly released [azure-functions-python-1.2.1 module](https://pypi.org/project/azure-functions/1.2.1/).
